src/Cli/dotnet/commands/dotnet-test/IPC/NamedPipeServer.cs

```diff
@@ -0,0 +1,305 @@
+#nullable disable
+#pragma warning restore IDE0240 // Remove redundant nullable directive
+
+using System.Buffers;
+using System.Globalization;
+using System.IO.Pipes;
+
+namespace Microsoft.DotNet.Tools.Test;
+
+internal sealed class NamedPipeServer : NamedPipeBase, IServer
+{
+    private readonly Func<IRequest, Task<IResponse>> _callback;
+    private readonly NamedPipeServerStream _namedPipeServerStream;
+    private readonly CancellationToken _cancellationToken;
+    private readonly MemoryStream _serializationBuffer = new();
+    private readonly MemoryStream _messageBuffer = new();
+    private readonly byte[] _readBuffer = new byte[250000];
+    private Task _loopTask;
+    private bool _disposed;
+
+    public NamedPipeServer(
+        string name,
+        Func<IRequest, Task<IResponse>> callback,
+        CancellationToken cancellationToken)
+        : this(GetPipeName(name), callback, cancellationToken)
+    {
+    }
+
+    public NamedPipeServer(
+        PipeNameDescription pipeNameDescription,
+        Func<IRequest, Task<IResponse>> callback,
+        CancellationToken cancellationToken)
+        : this(pipeNameDescription, callback, maxNumberOfServerInstances: 1, cancellationToken)
+    {
+    }
+
+    public NamedPipeServer(
+        PipeNameDescription pipeNameDescription,
+        Func<IRequest, Task<IResponse>> callback,
+        int maxNumberOfServerInstances,
+        CancellationToken cancellationToken)
+    {
+        _namedPipeServerStream = new((PipeName = pipeNameDescription).Name, PipeDirection.InOut, maxNumberOfServerInstances);
+        _callback = callback;
+        _cancellationToken = cancellationToken;
+    }
+
+    public PipeNameDescription PipeName { get; private set; }
+
+    public bool WasConnected { get; private set; }
+
+    public async Task WaitConnectionAsync(CancellationToken cancellationToken)
+    {
+        await _namedPipeServerStream.WaitForConnectionAsync(cancellationToken);
+        WasConnected = true;
+        _loopTask = Task.Run(
+            async () =>
+        {
+            try
+            {
+                await InternalLoopAsync(_cancellationToken);
+            }
+            catch (OperationCanceledException ex) when (ex.CancellationToken == _cancellationToken)
+            {
+                // We are being cancelled, so we don't need to wait anymore
+                return;
+            }
+            catch (Exception ex)
+            {
+                Environment.FailFast($"[NamedPipeServer] Unhandled exception:{Environment.NewLine}{ex}", ex);
+            }
+        }, cancellationToken);
+    }
+
+    /// <summary>
+    /// 4 bytes = message size
+    /// ------- Payload -------
+    /// 4 bytes = serializer id
+    /// x bytes = object buffer.
+    /// </summary>
+    private async Task InternalLoopAsync(CancellationToken cancellationToken)
+    {
+        int currentMessageSize = 0;
+        int missingBytesToReadOfWholeMessage = 0;
+        while (true)
```

Comment:
It probably doesn't change much but might be worth to do 

```suggestion
        while (!cancellationToken.IsCancellationRequested)
```